warp/Network/Wai/Handler/Warp/HTTP2/Types.hs

```diff
@@ -0,0 +1,235 @@
+{-# LANGUAGE OverloadedStrings, RecordWildCards, CPP #-}
+
+module Network.Wai.Handler.Warp.HTTP2.Types where
+
+import Blaze.ByteString.Builder (Builder)
+#if __GLASGOW_HASKELL__ < 709
+import Control.Applicative ((<$>),(<*>))
+#endif
+import Control.Concurrent
+import Control.Concurrent.STM
+import Control.Exception (SomeException)
+import Control.Monad (void)
+import Control.Reaper
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Builder.Extra as B
+import Data.IntMap.Strict (IntMap, IntMap)
+import qualified Data.IntMap.Strict as M
+import qualified Network.HTTP.Types as H
+import Network.Wai (Request, Response)
+import Network.Wai.Handler.Warp.IORef
+import Network.Wai.Handler.Warp.Types
+
+import Network.HTTP2
+import Network.HTTP2.Priority
+import Network.HPACK
+
+----------------------------------------------------------------
+
+http2ver :: H.HttpVersion
+http2ver = H.HttpVersion 2 0
+
+isHTTP2 :: Transport -> Bool
+isHTTP2 TCP = False
+isHTTP2 tls = useHTTP2
+  where
+    useHTTP2 = case tlsNegotiatedProtocol tls of
+        Nothing    -> False
+        Just proto -> "h2-" `BS.isPrefixOf` proto
+
+----------------------------------------------------------------
+
+data Input = Input Stream Request Priority
+
+----------------------------------------------------------------
+
+data Control a = CFinish
+               | CNext a
+               | CNone
+
+instance Show (Control a) where
+    show CFinish   = "CFinish"
+    show (CNext _) = "CNext"
+    show CNone     = "CNone"
+
+type DynaNext = WindowSize -> IO Next
+
+type BytesFilled = Int
+
+data Next = Next BytesFilled (Control DynaNext)
+
+data Output = OFinish
+            | OGoaway ByteString
+            | OFrame  ByteString
+            | OResponse Stream Response Aux
+            | ONext Stream DynaNext
+
+----------------------------------------------------------------
+
+data Sequence = SFinish
+              | SFlush
+              | SBuilder Builder
+
+data Sync = SyncNone
+          | SyncFinish
+          | SyncNext DynaNext
+
+data Aux = Oneshot Bool
+         | Persist (TBQueue Sequence) (TVar Sync)
+
+data Leftover = LZero
+              | LOne B.BufferWriter
+              | LTwo ByteString B.BufferWriter
+
+----------------------------------------------------------------
+
+data Context = Context {
+    http2settings      :: IORef Settings
+  , streamTable        :: StreamTable
+  , concurrency        :: IORef Int
+  , continued          :: IORef (Maybe StreamId)
```

Comment:
This seems to indicate the stream id that the next frame should be a continuation of?  Documenting this context record would make reading the code a lot easier I think.